### ISSUE TITLE
Message events

### ISSUE THREAD

**Issue by fnutt** [MultiChat/Development#28]
I'm putting forward a feature reuqest to make it easier for PurpleBungeeIRC to integrate into MultiChat. I've specifically asked the plugin author of PurpleBungeeIRC (cnaude) to add support for it, as the current situation is support for an outdated version of Herochat premium plugin as well as BungeeChat which PurpleBungeeIRC originally was going to support does not seem to care much about keeping the plugin up to date.

> 06:24 <fnutt> cnaude: Does Multichat have those events you requested on BungeeChat github?
> 06:24 <@cnaude> yes and no
> 06:24 <@cnaude> it's going to take a few days
> 06:24 <@cnaude> i might sent a pull request

This feature request will also make it much easier for other plugins to integrate with MultiChat.

**What I would like to see:**
Four different event messages - one for each message type (global, local, admin and moderator channels), or one message event with all of the information. Information which is needed are: Sender, Receiver, Channel name etc.

Maybe something like this:
- xyz.olivermartin.multichat.bungee.GlobalMessageEvent;
- xyz.olivermartin.multichat.bungee.LocalMessageEvent;
- xyz.olivermartin.multichat.bungee.AdminMessageEvent;
- xyz.olivermartin.multichat.bungee.ModeratorMessageEvent;

I hope for positive reply. Thank you in advance!

**Comment by OllieMartin** [comment]
Hi there,
Message events like you suggest are something I have been meaning to add for a while. I was going to wait for future updates with a lot of other changes, but it is something I could probably do a lot sooner (v1.6.1). 

I will have a look at what I can do, and can potentially work with PurpleBungeeIRC to try and find a solution for you. I haven’t used their plugin before but I will take a look.

So in summary, it’s something I’m willing to try and support!

I hope that sounds promising to you, thanks for the request.

~ Ollie

**Comment by fnutt** [comment]
Sounds great. Highly appreciated you give it some priority.

I've notified cnaude about your reply. If you want, please contact him on #purpleirc @ ipo.esper.net:6697

**Comment by cnaude** [comment]
:+1:

**Comment by OllieMartin** [comment]
Hi,

I just wanted to keep you updated with my progress. In the future versions like 1.7 I want to really use custom events as much as possible. So this is where we will really get a proper API that can be used easily etc.. However; I have attempted to implement into version 1.6.1 some "IRCEvents" which are much more simple events this plugin will fire, which I think can be used by IRC plugins like you named above. They are essentially read-only, so you can't modify the messages MultiChat will send by using them.

If you take a look at the repo https://github.com/MultiChat/Development/tree/version-1.6.1/multichat/src/main/java/xyz/olivermartin/multichat/bungee/events then you can see the kind of thing I have come up with.

Let me know what you think. I am trying to keep these events as simple as possible, because in future versions like 1.7 I want to implement the 'proper' events which could be used to actually alter the messages sent etc.

I also haven't tested these events yet at all, but I think they should work.

**Comment by cnaude** [comment]
Hi,

The events look like they would work. My only recommendation would to make the names a little more generic. Instead of MultiChatGlobalIRCEvent maybe GlobalChatEvent etc. If possible it would be nice to have access to a formatted and raw message. The raw messages are helpful for when devs want to do their own formatting.

**Comment by OllieMartin** [comment]
Hi,
I can definitely change the names, I only made them so specific as I thought it would prevent any conflicts, but on reflection it really shouldn’t make any difference!

And as for a “raw” message, I assume you mean with all the formatting codes stripped? So instead of “hello &6&lworld!”, the raw version would be “hello world”?

But yes I can easily accommodate that, and will let you know my progress :)

**Comment by OllieMartin** [comment]
> 
> 
> Hi,
> 
> The events look like they would work. My only recommendation would to make the names a little more generic. Instead of MultiChatGlobalIRCEvent maybe GlobalChatEvent etc. If possible it would be nice to have access to a formatted and raw message. The raw messages are helpful for when devs want to do their own formatting.

I've done some refactoring and added in a few methods. Take a look here if you get a chance https://github.com/MultiChat/Development/tree/version-1.6.1/multichat/src/main/java/xyz/olivermartin/multichat/bungee/events

I think this is a better way of doing it and hope it solves any concerns you had before :)

**Comment by cnaude** [comment]
> > Hi,
> > The events look like they would work. My only recommendation would to make the names a little more generic. Instead of MultiChatGlobalIRCEvent maybe GlobalChatEvent etc. If possible it would be nice to have access to a formatted and raw message. The raw messages are helpful for when devs want to do their own formatting.
> 
> I've done some refactoring and added in a few methods. Take a look here if you get a chance https://github.com/MultiChat/Development/tree/version-1.6.1/multichat/src/main/java/xyz/olivermartin/multichat/bungee/events
> 
> I think this is a better way of doing it and hope it solves any concerns you had before :)

Looks great. I think that will work.

**Comment by OllieMartin** [comment]
(The events discussed above were implemented in version 1.6.1 which has now been released on spigot)
